python/cuml/utils/input_utils.py

```diff
@@ -0,0 +1,192 @@
+    """
+    return cudf.bindings.cudf_cpp.get_column_data_ptr(col._column)
+
+
+def get_dtype(X):
+    """
+    Returns dtype of obj as a Numpy style dtype (like np.float32)
+    """
+    if isinstance(X, cudf.DataFrame):
+        dtype = np.dtype(X[X.columns[0]]._column.dtype)
+    elif (isinstance(X, cudf.Series)):
+        dtype = np.dtype(X._column.dtype)
+    elif isinstance(X, np.ndarray):
+        dtype = X.dtype
+    elif cuda.is_cuda_array(X):
+        dtype = X.dtype
+    elif cuda.devicearray.is_cuda_ndarray(X):
+        dtype = X.dtype
+
+    return dtype
+
+
+def input_to_dev_array(X, order='F', deepcopy=False,
+                       check_dtype=False, convert_to_dtype=False,
+                       check_cols=False, check_rows=False):
+    """
+    Convert input X to device array suitable for C++ methods
+    Acceptable input formats:
+    * cuDF Dataframe - returns a deep copy always
+    * cuDF Series - returns by reference or a deep copy depending on
+        `deepcopy`
+    * Numpy array - returns a copy in device always
+    * cuda array interface compliant array (like Cupy) - returns a
+        reference unless deepcopy=True
+    * numba device array - returns a reference unless deepcopy=True
+
+    Returns: namedtuple('dev_array', 'array pointer n_rows n_cols dtype')
+
+    `dev_array` is a new device array if the input was not a numba device
+        array. It is a reference to the input X if it was a numba device array
+        or cuda array interface compliant (like cupy)
+    """
+
+    if convert_to_dtype:
+        X = convert_dtype(X, to_dtype=convert_to_dtype)
+        check_dtype = False
+
+    if isinstance(X, cudf.DataFrame):
+        dtype = np.dtype(X[X.columns[0]]._column.dtype)
+        if order == 'F':
+            X_m = X.as_gpu_matrix(order='F')
+        elif order == 'C':
+            X_m = cuml.utils.numba_utils.row_matrix(X)
+
+    elif (isinstance(X, cudf.Series)):
+        if deepcopy:
+            X_m = X.to_gpu_array()
+        else:
+            X_m = X._column._data.mem
+
+    elif isinstance(X, np.ndarray):
+        dtype = X.dtype
+        X_m = rmm.to_device(np.array(X, order=order, copy=deepcopy))
+
+    elif cuda.is_cuda_array(X):
+        # Use cuda array interface to create a device array by reference
+        X_m = cuda.as_cuda_array(X)
+        if deepcopy:
+            out_dev_array = rmm.device_array_like(X_m)
+            out_dev_array.copy_to_device(X_m)
+            X_m = out_dev_array
+
+    elif cuda.devicearray.is_cuda_ndarray(X):
+        if deepcopy:
+            out_dev_array = rmm.device_array_like(X)
+            out_dev_array.copy_to_device(X)
+            X_m = out_dev_array
+        else:
+            X_m = X
+
+    else:
+        msg = "X matrix format " + str(X.__class__) + " not supported"
+        raise TypeError(msg)
+
+    dtype = X_m.dtype
+
+    if check_dtype:
+        if dtype != check_dtype:
+            del X_m
+            raise TypeError("ba")
```

Comment:
Should there be a more descriptive exception message?

Reply:
lol forgot to replace this 2